### PR TITLE
Optimized API call, query timing for the table aws_ec2_instance_type and added support to accept the wildcard pattern(`t2*`) for instance type. Closes #2292

### LINKED ISSUE #2292: aws_ec2_instance_type takes over 4 minutes
**Describe the bug**
A Query on the instance type table aws_ec2_instance_type took over 4 minutes to be finished.

**Steampipe version (`steampipe -v`)**
Steampipe v0.23.1

**Plugin version (`steampipe plugin list`)**

+--------------------------------------------+---------+----------------------------------------------------------------------+
| Installed                                  | Version | Connections                                                          |
+--------------------------------------------+---------+----------------------------------------------------------------------+
| hub.steampipe.io/plugins/turbot/aws@latest | 0.145.0 | .....                                                                |
+--------------------------------------------+---------+----------------------------------------------------------------------+

**To reproduce**
select * from dsp.aws_ec2_instance_type where instance_type like 't2%';

**Expected behavior**

A shorter time for finishing the query, like 40s max.

**Additional context**

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>

```
N/A
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
> select instance_type, instance_type_pattern from aws_ec2_instance_type where instance_type_pattern = 'c7i.*'
+----------------+-----------------------+
| instance_type  | instance_type_pattern |
+----------------+-----------------------+
| c7i.16xlarge   | c7i.*                 |
| c7i.xlarge     | c7i.*                 |
| c7i.large      | c7i.*                 |
| c7i.metal-48xl | c7i.*                 |
| c7i.4xlarge    | c7i.*                 |
| c7i.2xlarge    | c7i.*                 |
```
</details>

### CODE REVIEW

**Comment by ParthaI** [comment]
### Changes Overview:

#### API Call Optimization:
- The table now makes **2 API calls per query**, compared to earlier where it made **API calls for each row**.
- Query execution time has been reduced by **~75%**.

---

#### Performance Comparison:

**Earlier (73.8s):**
```sql
> select count(*) from aws_ec2_instance_type;
+-------+
| count |
+-------+
| 11403 |
+-------+

Time: 73.8s. Rows returned: 1. Rows fetched: 11,380. Hydrate calls: 11,403.
```

**Now (12.2s):**
```sql
> select count(*) from aws_ec2_instance_type;
+-------+
| count |
+-------+
| 11403 |
+-------+

Time: 12.2s. Rows returned: 1. Rows fetched: 11,403. Hydrate calls: 0.
```
---

#### Wildcard Search Support:
- Added a new column: **`instance_type_pattern`** to enable wildcard pattern searches.
- In the query `select * from aws_ec2_instance_type where instance_type = 't2*'`, the API returned empty rows due to PostgreSQL-level filtering.
    - The `instance_type` column contains exact values like `'t2.small'`, not `'t2*'`, causing a mismatch between the column value and the wildcard pattern.
    - To address this, the **`instance_type_pattern`** column was introduced to allow proper filtering using wildcard patterns.

#### Example Query:
```sql
> select instance_type, instance_type_pattern from aws_ec2_instance_type where instance_type_pattern = 'c7i.*';
+----------------+-----------------------+
| instance_type  | instance_type_pattern |
+----------------+-----------------------+
| c7i.16xlarge   | c7i.*                 |
| c7i.xlarge     | c7i.*                 |
| c7i.large      | c7i.*                 |
| c7i.metal-48xl | c7i.*                 |
| c7i.4xlarge    | c7i.*                 |
| c7i.2xlarge    | c7i.*                 |
| c7i.metal-24xl | c7i.*                 |
| c7i.8xlarge    | c7i.*                 |
| c7i.12xlarge   | c7i.*                 |


Time: 1.3s. Rows returned: 176. Rows fetched: 176. Hydrate calls: 0.

Scans:
  1) aws_ec2_instance_type.aws: Time: 1.2s. Fetched: 176. Hydrates: 0. Quals: instance_type_pattern = 'c7i.*'.
```

The above query fetched and displayed **176 rows** in just **1.3 seconds**.

---

@e-gineer / @misraved, your feedback on the table design and suggestions regarding the column naming for **`instance_type_pattern`** would be greatly appreciated.

Thank you!

File: aws/table_aws_ec2_instance_type.go [hunk 1]
```diff
@@ -272,99 +241,153 @@ func listAwsInstanceTypesOfferings(ctx context.Context, d *plugin.QueryData, h *
...
+	return instanceTypes, nil
+}
 
-			// Context can be cancelled due to manual cancellation or the limit has been hit
-			if d.RowsRemaining(ctx) == 0 {
-				return nil, nil
-			}
+// Helper function to filter instance types by a pattern like t2-*, m5-*, etc.
+func filterInstanceTypesByPattern(ctx context.Context, instanceTypes []types.InstanceType, pattern string) []types.InstanceType {
+	plugin.Logger(ctx).Error("Pattern =>>>", pattern)
```
Comment (misraved) [inline aws/table_aws_ec2_instance_type.go]:
```suggestion

```

File: aws/table_aws_ec2_instance_type.go [hunk 2]
```diff
@@ -272,99 +241,153 @@ func listAwsInstanceTypesOfferings(ctx context.Context, d *plugin.QueryData, h *
...
+	// - "c7i-*" matches (e.g., c7i-flex.2xlarge).
+	pattern = strings.ReplaceAll(pattern, ".", "\\.")
+	pattern = strings.ReplaceAll(pattern, "*", ".+")
+	var matchedInstanceTypes []types.InstanceType
+	re := regexp.MustCompile(pattern)
+
+	for _, instanceType := range instanceTypes {
+
+		if re.MatchString(string(instanceType)) {
+			plugin.Logger(ctx).Error("Pattern matched with =>>>", instanceType)
```
Comment (misraved) [inline aws/table_aws_ec2_instance_type.go:322]:
```suggestion

```

**Review by misraved — COMMENTED** [review]
@ParthaI what happens if we pass `limit 15` to the query? In an ideal case scenario, I wouldn't want to make a batch API call, that passes in 100 instance types.

Please take a look at the minor review comments, thanks!!

**Comment by ParthaI** [comment]
@misraved, thank you for the suggestion! I’ve updated the logic to optimize the API call using the `limit` query parameter.

**Pull request merged**